source/Octopus.Tentacle/Kubernetes/KubernetesEventMonitor.cs

```diff
@@ -0,0 +1,84 @@
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using k8s.Models;
+using Octopus.Tentacle.Kubernetes.Diagnostics;
+
+namespace Octopus.Tentacle.Kubernetes
+{
+    public interface IKubernetesEventMonitor
+    {
+        Task CacheNewEvents(CancellationToken cancellationToken);
+    }
+
+    public class KubernetesEventMonitor : IKubernetesEventMonitor
+    {
+        readonly IKubernetesAgentMetrics agentMetrics;
+        readonly IKubernetesEventService eventService;
+
+        public KubernetesEventMonitor(IKubernetesAgentMetrics agentMetrics, IKubernetesEventService eventService)
+        {
+            this.agentMetrics = agentMetrics;
+            this.eventService = eventService;
+        }
+
+        public async Task CacheNewEvents(CancellationToken cancellationToken)
+        {
+            var allEvents = await eventService.FetchAllEventsAsync(GetNamespace(), cancellationToken) ?? new Corev1EventList(new List<Corev1Event>());
+
+            var lastCachedEventTimeStamp = agentMetrics.GetLatestEventTimestamp();
+
+            var unseenEvents = allEvents.Items.Where(e =>
+            {
+                var eventTimestamp = GetLatestTimestampInEvent(e);
+                return eventTimestamp.HasValue && eventTimestamp.Value.ToUniversalTime() > lastCachedEventTimeStamp;
+            });
+            
+            foreach (var kEvent in unseenEvents)
+            {
+                if (IsRelevantForMetrics(kEvent))
+                {
+                    var eventTimestamp = GetLatestTimestampInEvent(kEvent)!.Value.ToUniversalTime();
+                    agentMetrics.TrackEvent(kEvent.Reason, kEvent.Name(), eventTimestamp);
+                }
+            }
+        }
+
+        bool IsRelevantForMetrics(Corev1Event kubernetesEvent)
+        {
+            return IsNfsPodRestart(kubernetesEvent) || IsPodRestartDueToStaleNfs(kubernetesEvent);
```

Comment:
Do we also want to track Tentacle pod restarts? That can help us with issues that aren't necessarily NFS related.

Reply:
yeah - that's probably smart.